[memory profiling] add a facility to gather combined C++/Python/TorchScript stack traces.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95753
* __->__ #95541


    This refactors the stack trace facility specific to memory profiling
    in python+cuda to make a generic facility to generate combined stack
    traces.
    
    The generic facility (combined_traceback.h) does not require
    python to be around to work, but will return python stacks if it is
    present.
    
    This facility is then used to add support for stack trace gathering in memory profiling that
    happens directly from C++.
    
    It is also used to expose a python API for gathering and symbolizing
    combineds stacks.

zdevito committed:
[memory profiling] add stack traces to C++ API

This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add stack traces to C++ API"


This adds support for stack trace gathering in memory profiling that
happens directly from C++. It records stacks with both TorchScript and
C++ frames.

[ghstack-poisoned]

zdevito commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -596,186 +595,14 @@ PyObject* THCPModule_resetPeakMemoryStats(PyObject* _unused, PyObject* arg) {
   Py_RETURN_NONE;
 }
 
-struct Frame {
```
This code is moving to `combined_traceback.cpp`. The change in logic is only to: * remove pybind11 specific code, and * deduplicate repeat Python stack frames to avoid creating additional python objects.

ezyang commented on torch/csrc/profiler/python/init.cpp:
```diff
@@ -292,6 +293,12 @@ void initPythonBindings(PyObject* module) {
+      m, "CapturedTraceback");
+  m.def("gather_traceback", CapturedTraceback::gather);
+  m.def("symbolize_tracebacks", py_symbolize);
+  installCapturedTracebackPython();
```
I'd definitely appreciate pyi type stubs for tehse bindings at the appropriate spot (maybe _C/__init__.pyi.in?)

ezyang replied:
And docs!

ezyang commented on torch/csrc/profiler/python/combined_traceback.h:
```diff
@@ -0,0 +1,14 @@
+
+// symbolize combined traceback objects, converting them into lists of
+// dictionaries that are easily consumed in python.
+std::vector<pybind11::object> py_symbolize(
```
vector return of py::object 🤔

ezyang commented on torch/csrc/cuda/memory_snapshot.h:
```diff
@@ -10,7 +10,11 @@ namespace cuda {
+    bool record_context = true,
+    int64_t trace_alloc_max_entries = 1,
+    bool trace_alloc_record_context = false,
+    bool record_cpp_context = false);
```
hella boolean blindness here

zdevito replied:
It just matches what the internal Python API is today. I am going to clean it up once I stop adding features and I can bin them into a couple sane options.

ezyang commented on torch/csrc/profiler/combined_traceback.h:
```diff
@@ -0,0 +1,55 @@
+  // index into all_frames, so that
+  // it is possible to dedupe frame objects in
+  // construction of python objects
+  std::vector<std::vector<uint64_t>> tracebacks;
```
what exactly is the nested vector here?

ezyang replied:
oh, it's multiple tracebacks, and then each traceback is indexes into all frames

ezyang commented on torch/csrc/profiler/combined_traceback.h:
```diff
@@ -0,0 +1,55 @@
+  std::vector<jit::StackEntry> script_frames_;
+  friend SymbolizedTracebacks symbolize(
+      const std::vector<CapturedTraceback*>& to_symbolize);
+  Python* python_ = nullptr;
```
Owning? Non-owning?

zdevito committed:
Update on "[memory profiling] add a facility to gather combined C++/Python/TorchScript stack traces."



    This refactors the stack trace facility specific to memory profiling
    in python+cuda to make a generic facility to generate combined stack
    traces.
    
    The generic facility (combined_traceback.h) does not require
    python to be around to work, but will return python stacks if it is
    present.
    
    This facility is then used to add support for stack trace gathering in memory profiling that
    happens directly from C++.
    
    It is also used to expose a python API for gathering and symbolizing
    combineds stacks.
    

[ghstack-poisoned]

ezyang commented on torch/csrc/profiler/combined_traceback.h:
```diff
@@ -0,0 +1,64 @@
+ private:
+  std::vector<PyFrame> frames_;
+  std::vector<void*> cpp_frames_;
+  std::vector<jit::StackEntry> script_frames_;
```
one thing I wonder just reading the header here is how you know which order to reinterleave them. You told me about how to do this in person but I forgot.

zdevito replied:
It can't be interleaved without knowing how the CPP code is symbolized, so the implementation of `symbolize` is where it lives.

ezyang commented on torch/csrc/profiler/python/combined_traceback.cpp:
```diff
@@ -0,0 +1,123 @@
+  void appendSymbolized(
+      const std::vector<CapturedTraceback::PyFrame>& to_symbolize,
+      SymbolizedTracebacks& result) override {
+    auto torch = py::module::import("torch");
```
hmm?

zdevito replied:
oh, that is suppose to be part of the next PR, split them wrong.

ezyang commented on torch/csrc/profiler/python/combined_traceback.cpp:
```diff
@@ -0,0 +1,123 @@
+      auto f_back = PyFrame_GetBack(f);
+      Py_XDECREF(f);
+      f = f_back;
+    }
```
A little confused: does f leak here?

zdevito replied:
I think it does, good catch! That has been in there for awhile but has probably not caused any problems because it is leaking the root frame (which is only going to go dead when python exits for the most part).

zdevito replied:
Oh, never mind `f` must be null at the exit to the loop.

ezyang reviewed (approved):
see comments

zdevito committed:
Update on "[memory profiling] add a facility to gather combined C++/Python/TorchScript stack traces."



    This refactors the stack trace facility specific to memory profiling
    in python+cuda to make a generic facility to generate combined stack
    traces.
    
    The generic facility (combined_traceback.h) does not require
    python to be around to work, but will return python stacks if it is
    present.
    
    This facility is then used to add support for stack trace gathering in memory profiling that
    happens directly from C++.
    
    It is also used to expose a python API for gathering and symbolizing
    combineds stacks.
    

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] add a facility to gather combined C++/Python/TorchScript stack traces."



    This refactors the stack trace facility specific to memory profiling
    in python+cuda to make a generic facility to generate combined stack
    traces.
    
    The generic facility (combined_traceback.h) does not require
    python to be around to work, but will return python stacks if it is
    present.
    
    This facility is then used to add support for stack trace gathering in memory profiling that
    happens directly from C++.
    
    It is also used to expose a python API for gathering and symbolizing
    combineds stacks.
    

[ghstack-poisoned]

zdevito commented:
ok, I think I added comments or replied to each comment. Let's see if this compiles through windows and oddball pythons...

zdevito committed:
Update on "[memory profiling] add a facility to gather combined C++/Python/TorchScript stack traces."



    This refactors the stack trace facility specific to memory profiling
    in python+cuda to make a generic facility to generate combined stack
    traces.
    
    The generic facility (combined_traceback.h) does not require
    python to be around to work, but will return python stacks if it is
    present.
    
    This facility is then used to add support for stack trace gathering in memory profiling that
    happens directly from C++.
    
    It is also used to expose a python API for gathering and symbolizing
    combineds stacks.
    

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst
